mapproxy/test/system/test_cache_geopackage.py

```diff
@@ -0,0 +1,309 @@
+# This file is part of the MapProxy project.
+# Copyright (C) 2011 Omniscale <http://omniscale.de>
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import with_statement, division
+
+import os
+import shutil
+
+from io import BytesIO
+
+from mapproxy.request.wms import WMS111MapRequest
+from mapproxy.test.http import MockServ
+from mapproxy.test.image import is_png, create_tmp_image
+from mapproxy.test.system import prepare_env, create_app, module_teardown, SystemTest
+from mapproxy.cache.tile import Tile
+from mapproxy.test.unit.test_cache_tile import TileCacheTestBase
+from mapproxy.cache.geopackage import GeopackageCache, GeopackageLevelCache
+from mapproxy.config.loader import load_configuration
+from mapproxy.image import ImageSource
+import time, threading
+from nose.tools import eq_
+import sqlite3
+
+test_config = {}
+
+
+def setup_module():
+    prepare_env(test_config, 'cache_geopackage.yaml')
+
+    shutil.copy(os.path.join(test_config['fixture_dir'], 'cache.gpkg'),
+        test_config['base_dir'])
+    create_app(test_config)
+
+def teardown_module():
+    module_teardown(test_config)
+
+
+class TestGeopackageCache(SystemTest, TileCacheTestBase):
+    config = test_config
+    table_name = 'cache'
+
+    def setup(self):
+        configuration = load_configuration(self.config.get('config_file'))
+        TileCacheTestBase.setup(self)
+        self.cache = GeopackageCache(os.path.join(self.cache_dir, 'tmp.geopackage'),
+                                     configuration.grids.get('GLOBAL_GEODETIC'),
+                                     self.table_name)
+
+    def teardown(self):
+        if self.cache:
+            self.cache.cleanup()
+        TileCacheTestBase.teardown(self)
+
+    def test_get_map_cached(self):
+        prepare_env(test_config, 'cache_geopackage.yaml')
```

Comment:
`prepare_enc` and `create_app` are "expensive" and should only be called on module or class level. They should also always be followed by a `teardown` call.

Normally, for the system tests there is one Python test file for each YAML configuration. You can work around that by creating a `SystemTest` subclass for each test configuration and add a `setup_class` and `teardown_class` methods that sets everything up. See http://nose.readthedocs.io/en/latest/writing_tests.html#test-classes

But see also my next comment. Maybe multiple classes are not necessary.